pkg/health/checker/checker_test.go

```diff
@@ -52,66 +51,60 @@ func TestService(t *testing.T) {
 	Assert(t).AreEqual(results["node1"], expected, "Unexpected results calling Service()")
 }
 
-func TestWatchHealth(t *testing.T) {
-	fakeKV := kptest.NewFakeKV()
-	healthChecker := &consulHealthChecker{
-		kv: fakeKV,
-	}
+func TestPublishHealth(t *testing.T) {
+	// This channel imitates the channel that consulutil.WatchPrefix would return
+	healthListChan := make(chan api.KVPairs)
+	quitCh := make(chan struct{})
+	outCh := make(chan []*health.Result, 1)
+	defer close(outCh)
+	defer close(quitCh)
 
-	retChan := make(chan []*health.Result)
-	errChan := make(chan error)
-	quitChan := make(chan struct{})
+	errCh := publishLatestHealth(healthListChan, quitCh, outCh)
 
-	dummyHealthResult := &health.Result{
-		ID:      "ID",
-		Node:    "node1.example.com",
-		Service: "Service",
-		Status:  "passing",
-		Output:  "output",
+	go func() {
+		err, open := <-errCh
+		if err != nil {
+			t.Fatal(err)
+		}
+		if !open {
+			return
+		}
+	}()
+
+	oldStatus := health.HealthState("passing")
+	newStatus := health.HealthState("critical")
+	hrOld := &health.Result{
+		Status: oldStatus,
 	}
-	dummyBuf, err := json.Marshal(dummyHealthResult)
+	hrOldJSON, err := json.Marshal(hrOld)
 	if err != nil {
-		t.Fatalf("json marshal err: %v", err)
+		t.Fatal("json marshal err: %v", err)
 	}
+	oldKV := &api.KVPair{Key: "health/service/node1.example.com", Value: hrOldJSON}
 
-	_, _, err = fakeKV.CAS(&api.KVPair{Key: "health/service/node1.example.com", Value: dummyBuf}, nil)
+	hrNew := &health.Result{
+		Status: newStatus,
+	}
+	hrNewJSON, err := json.Marshal(hrNew)
 	if err != nil {
-		t.Fatalf("error: %v", err)
+		t.Fatal("json marshal err: %v", err)
 	}
+	newKV := &api.KVPair{Key: "health/service/node1.example.com", Value: hrNewJSON}
 
-	go func() {
-		defer func() {
-			quitChan <- struct{}{}
-		}()
-		for {
-			select {
-			case result := <-retChan:
-				res := result[0]
-				if res.ID != "ID" {
-					t.Fatalf("Expected ID to match")
-				}
+	// Basic test that publishLatestHealth drains the channels correctly
+	healthListChan <- api.KVPairs{oldKV}
```

Comment:
nit: `select {}` on a `<-time.After` for these so the test doesn't hang forever if someone introduces a bug